docs/01-tooling/03-fast-and-isolated-JS-unit-tests.md

```diff
@@ -2,10 +2,159 @@
 
 ### Intro
 
+For testing frontend code, MediaWiki provides a browser based QUnit setup. For
+running the tests, you have to spin up MediaWiki, and load
+Special:JavaScriptTest in your browser, which will run all the QUnit tests for
+all the extensions and MediaWiki itself.
+
+From then on, you can filter by module or string, hide passed tests, and a few
+other things like select to load the assets in production mode or development
+mode (`debug=true`).
+
+Like any testing strategy, this setup comes with tradeoffs. Specifically, there
+are a couple that we have had problems with:
+
+#### Tests usually take a long time to run
+
+With this setup, they have to go through to the MediaWiki server,
+ResourceLoader, and then run in the browser. This is specially noticeable in
+development mode, which we often enable to get readable stack traces and error
+messages on test failures, but makes the test run take a lot longer.
+
+As a result of this, writing tests is costlier than it should, which discourages
+developers to write and run tests, and over time ends up affecting the quality
+of our code and QUnit test suites. It also puts significant barriers to TDD
+style workflows, which rely on constantly running tests and require a fast
+feedback cycle with the developer.
+
+#### It is very hard to write isolated unit tests
+
+In this environment, the real MediaWiki, global state, and the browser
+environment are available. As a result, tests are written often as integration
+tests, relying on implicit behavior, modules and state being available to
+perform the testing.
+
+This is not a problem by itself, integration tests are important and have very
+valid use cases and a place in the testing process, but this environment itself
+makes it extremely complicated to write isolated unit tests, because of all the
+global state, variables, and existing loaded code.
+
+As a result, tests written end up coupled to implicit behavior and state, which
+makes them very brittle or overly verbose because of the extensive mocking, and
+most of them are big integration tests, which makes them slower to run. All of
+this adds up to the previous point, making it an even slower moving setup for
+writing tests, with the same outcomes.
+
 ### Requirements
 
-### Solutions
+We need a way to write tests for our frontend JavaScript that:
+
+* Encourage and enforce isolation
+  * Without global state or a full MediaWiki environment running
+* Start up and run the tests very fast
+* Re-runs tests when our source files change automatically, without having to
+  wait for the developer to go to the browser and run the tests
+
```

Comment:
I would also add a bullet to the effect of "indicate clearly when a failure occurs and where it is." I'm currently working in Popups and have lost a lot of time because of how errors are either slurped up by tap-dot or flood the screen with tap's verbosity.

Reply:
Right, I totally agree. I've added it.

The fault in Popups specifically is of the implementation, which I'm going to fix soon, given QUnit has released 2.6.0 with the --require option! https://github.com/qunitjs/qunit/releases

After I fix it I'll amend the last part of the post to remove the future improvements part.